### PR TITLE
Adjust filters *after* init

### LINKED ISSUE #7: Adjust filters *after* init
Add a feature to adjust the log filters *after* `init` is called. See [flexi_logger](https://docs.rs/flexi_logger/0.24.2/flexi_logger/struct.LoggerHandle.html) as an example. 

The use case is log filter stored on e.g `/data` on an Android system that may become availed *after* a `android_logd_logger* instance may need to be initialised. The final filter is not available when the logger is initialised and needs to be updated later.

Two possible approaches:

Either [Builder::init](https://docs.rs/android-logd-logger/0.3.2/android_logd_logger/struct.Builder.html#method.init) returns a handle that implements almost the same as `Builder`. The internally stored filter is updated and used accordingly. This handle must take a `Builder` as argument. This requires some sync between the handle and the longer implementation which is stored somewhere statically. Probably via `RwLock`. Downside here is the lock access upon each log entry.

Another approach is to provide a `update_filter` fn that basically does the same as [parse_filter](https://docs.rs/android-logd-logger/0.3.2/android_logd_logger/struct.Builder.html#method.parse_filters) and communicates the updates filter via a `log::debug!` to the logger implementation - the filter string is stored in `Metadata` and detected in the stream of logs in the Logger implementation. The advantage of this approach is that it's lock free. Downside is that it's less flexible than the handle approach above.

### PR DESCRIPTION
**Pull request by rasnaut**
- add log_configuration module
- add functionality for changing configuration settings

Fix #7

### CODE REVIEW

File: src/log_configuration.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
+use super::{Buffer, TagMode};
+
+use env_logger::filter::Filter;
+
+/// Documentation here
+pub struct LogConfiguration {
```
Comment (flxo) [inline src/log_configuration.rs]: This is the main API exposed for setting parameters. Can we name this just `Logger`?

File: src/log_configuration.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    pub(crate) prepend_module: bool,
+    #[allow(unused)]
+    pub(crate) pstore: bool,
+    pub(crate) buffer_id: Option<Buffer>,
+}
+
+impl LogConfiguration {
+    /// Initializes the Log Configuration
+    ///
+    ///
```
Comment (flxo) [inline src/log_configuration.rs]: Please remove the extra empty comment lines.

File: src/log_configuration.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    pub(crate) fn new(filter: Filter, tag: TagMode, prepend_module: bool, pstore: bool, buffer_id: Option<Buffer>) -> Self {
+        //let tag = TagMode::default();
```
Comment (flxo) [inline src/log_configuration.rs]: Leftover?

File: src/log_configuration.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            filter,
+            tag,
+            prepend_module,
+            pstore,
+            buffer_id,
+        }
+    }
+
+    /// Get Log configuration buffer field
+    ///
```
Comment (flxo) [inline src/log_configuration.rs]: Please remove the extra empty comment line.

File: src/log_configuration.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    /// Get Log configuration filter level
+    ///
+    pub fn get_level_filter(&self) -> log::LevelFilter {
+        self.filter.filter().to_level().unwrap().to_level_filter()
+    }
+
+    /// Set Log configuration buffer field
+    ///
```
Comment (flxo) [inline src/log_configuration.rs]: Please remove the empty extra line.

File: src/log_configuration.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        self.prepend_module
+    }
+
+    /// Get Log configuration pstore value
+    pub fn get_pstore(&self) -> bool {
+        self.pstore
+    }
+
+    /// Get Log configuration filter level
+    ///
```
Comment (flxo) [inline src/log_configuration.rs]: Please remove the extra empty line here.

File: src/log_configuration.rs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    /// Set Log configuration buffer field
+    ///
+    pub fn set_buffer(&mut self, buffer: Buffer) -> &mut Self {
+        self.buffer_id = Some(buffer);
+        self
+    }
+
+    /// Set Log configuration tag to custom
+    ///
```
Comment (flxo) [inline src/log_configuration.rs]: Please remove *all* empty lines in the comment below.

File: src/log_configuration.rs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    pub fn set_level_filter(&mut self, level_filter: log::LevelFilter) -> &mut Self {
+        let mut filter_builder = env_logger::filter::Builder::default();
+        let filter = filter_builder.filter_level(level_filter).build();
+        self.set_filter(filter);
+        self
+    }
+
+    /// Set Log configuration module and level filter
+    ///
+    pub fn set_mudule_and_level_filter(&mut self, new_module: &str, new_level_filter: log::LevelFilter) -> &mut Self {
```
Comment (flxo) [inline src/log_configuration.rs]: Here's a typo `mudule`.

File: src/logger.rs [hunk 9]
```diff
@@ -1,33 +1,89 @@
...
-    filter: Filter,
-    tag: TagMode,
-    prepend_module: bool,
-    #[allow(unused)]
-    pstore: bool,
-    #[allow(unused)]
-    buffer_id: Buffer,
+///Logger configuration handler stores access to logger configuration parameters
+///
+///
```
Comment (flxo) [inline src/logger.rs]: Again empty lines.

File: src/logger.rs [hunk 10]
```diff
@@ -1,33 +1,89 @@
...
+    /// # use log::LevelFilter;
+    /// # use android_logd_logger::Builder;
+    ///
+    /// let config = android_logd_logger::builder()
+    /// .parse_filters("debug")
+    /// .init();
+    ///
+    /// config.setter().set_level_filter(LevelFilter::Error);
+    /// ```
+    pub fn setter(&self) -> std::sync::RwLockWriteGuard<LogConfiguration> {
```
Comment (flxo) [inline src/logger.rs]: The synchronisation should be hidden in `Logger` and the lock guard should no be here. Why is there the `setter` call at all. To my understanding the instance returned from `init` is the "setter".

File: src/logger.rs [hunk 11]
```diff
@@ -1,33 +1,89 @@
...
+    ///
+    /// let config = android_logd_logger::builder()
+    /// .parse_filters("debug")
+    /// .init();
+    ///
+    /// let prepend_module = config.getter().get_prepend_module();
+    /// ```
+    pub fn getter(&self) -> std::sync::RwLockReadGuard<LogConfiguration> {
+        self.configuration_handler.read().unwrap()
+    }
```
Comment (flxo) [inline src/logger.rs]: Not sure if we need "getter" at all.

File: src/logger.rs [hunk 12]
```diff
@@ -1,33 +1,89 @@
...
+
+    pub(crate) fn get_config(&self) -> Arc<RwLock<LogConfiguration>> {
+        Arc::clone(&self.configuration_handler)
+    }
+}
+pub(crate) struct LoggerImpl {
+    configuration_handle: LoggerConfigHandler,
```
Comment (flxo) [inline src/logger.rs]: Please rename to `logger: Logger`.

File: src/lib.rs [hunk 13]
```diff
@@ -290,7 +291,7 @@ impl Builder {
     ///
     /// let mut builder = Builder::new();
     /// builder.buffer(Buffer::Crash)
-    ///     .init();
+    ///        .init();
```
Comment (flxo) [inline src/lib.rs]: Extra spaces?

File: src/lib.rs [hunk 14]
```diff
@@ -448,12 +449,14 @@ impl Builder {
     ///
     /// This function will fail if it is called more than once, or if another
     /// library has already initialized a global logger.
-    pub fn try_init(&mut self) -> Result<(), SetLoggerError> {
-        let logger = self.build();
+    pub fn try_init(&mut self) -> Result<logger::LoggerConfigHandler, SetLoggerError> {
```
Comment (flxo) [inline src/lib.rs]: This fn should return `Logger`.

File: src/log_configuration.rs [hunk 15]
```diff
@@ -0,0 +1,118 @@
+use super::{Buffer, TagMode};
+
+use env_logger::filter::Filter;
+
+/// Documentation here
```
Comment (flxo) [inline src/log_configuration.rs]: Missing docs.

**Review by flxo — CHANGES_REQUESTED** [review]
Hi. Thanks for the PR. 

Think we need to adjust the public API part.

File: src/log_configuration.rs (hunk 1 above)
Comment (rasnaut) [reply src/log_configuration.rs]: For my opinion name "Logger"  does not fully reflect the meaning of this class, or?

File: src/logger.rs (hunk 11 above)
Comment (rasnaut) [reply src/logger.rs]: Implementation LoggerImpl for Log contain a lot of places where need read fields, and for my opinion getter is most useful tool in this case

**Pull request merged**